lfs/batcher.go

```diff
@@ -6,36 +6,43 @@ import "sync/atomic"
 // be added to the batcher from multiple goroutines and pulled off in groups
 // when one of the following conditions occurs:
 //   * The batch size is reached
+//   * Flush() is called, forcing the batch to be returned immediately, as-is
 //   * Exit() is called
-// When an Exit() occurs, the group may be smaller than the batch size.
+// When an Exit() or Flush() occurs, the group may be smaller than the batch
+// size.
 type Batcher struct {
 	exited     uint32
 	batchSize  int
 	input      chan interface{}
 	batchReady chan []interface{}
+	flush      chan interface{}
 }
 
 // NewBatcher creates a Batcher with the batchSize.
 func NewBatcher(batchSize int) *Batcher {
 	b := &Batcher{
 		batchSize:  batchSize,
-		input:      make(chan interface{}, batchSize),
+		input:      make(chan interface{}),
```

Comment:
I think this is an interesting thought. The reason I un-buffered the channel in https://github.com/github/git-lfs/pull/1528/commits/98873d87c4b06beecbba57221e70dd975ce7d7d4 and https://github.com/github/git-lfs/pull/1528/commits/a27d92d1a48f5d6d57218501b60f00ebbbf75826 was so that I could ensure that after calling `Add()` a number of times in test, that they would _for sure_ be in the batch so that after `Flush()`-ing, I'd get them out.

However, I'm not actually sure we ever needed the buffered behavior in the first place. I could be wrong, since I haven't been in this part of the code in a while, but I ran some benchmarks to get a second opinion.

With buffering (on master):
```
~/g/g/lfs (master!) $ git rev-parse HEAD
b7b1f2b7985c8f7a18c563c70898f669ffb8c13c

~/g/g/lfs (master!) $ go test -run=BenchmarkAddingToBatcher -bench=./lfs
BenchmarkAddingToBatcher-4      10000000               151 ns/op
BenchmarkLsTreeParser-4           500000              4288 ns/op   # <- ignore
PASS
ok      github.com/github/git-lfs/lfs   3.880s
```

Without buffering (on 49489c0f43011c5f0638eb2fd17b993e7488a0a6):
```
~/g/g/lfs (batcher-truncate!) $ git rev-parse HEAD                                
49489c0f43011c5f0638eb2fd17b993e7488a0a6

~/g/g/lfs (batcher-truncate!) $ go test -run=BenchmarkAddingToBatcher -bench=./lfs
BenchmarkAddingToBatcher-4       3000000               551 ns/op
BenchmarkLsTreeParser-4           500000              4245 ns/op   # <- ignore
PASS
ok      github.com/github/git-lfs/lfs   4.422s
```

My initial thought after seeing these numbers was to buffer the channel. For future reference, one of the ways we could go about doing this is:

Upon receiving a `<-b.flush` signal, spin until we have completely drained the input channel:
```go
case <-b.input:
for {
        select {
                case t, ok := <-b.input:
                // Do the thing that we're doing above :-(
                default:
                // Channel is empty!
                break Acc
        }
}
```

However, @rubyist made a great point in that given that we're not adding to this batcher from multiple goroutines currently in LFS, having this sort of contention isn't a big deal. He also pointed out that since we're using this code to do network transfers, an extra 400ns/op isn't a huge deal either.

I think given our current use-case, and the additional complexity necessary to buffer the input channel, we should leave this as-is until we need to change it.

--

For reference, the benchmark I ran is in [this gist](https://gist.github.com/ttaylorr/dabc204dce7f0f808a1f69d9857229cd).